utils/estimated_histogram.hh

```diff
@@ -47,13 +47,302 @@
+ *
+ * lower bits are based on the high 2 MSB (ignoring the leading 1).
+
+ * for example: 330 (101001010)
+ * higher bits: log2(330/128) = 1
+ * low bit MSB: 330 = 01 (the lower two bits out of the upper 3)
+ * So the index: 101 = 5
+ *
+ *
+ * About the min/max and number of buckets.
+ * ========================================
+ *
+ * For MIN and MAX choose numbers that are a power of 2.
+ * The number of buckets will determine the resolution that should also be a power of 2
+ * So the total number of bucket should be log2(MAX/MIN) * Resolution + 1
+ *
+ * Limitation: You must set the MIN value to be higher then the resolution.
+ * For example, for a 2 bits resolution MIN should be 2^2 = 4 or higher.
+ *
+ */
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+requires (Min > 0 && Min < Max)
+class approx_exponential_histogram {
+    std::array<uint64_t, NumBuckets> _buckets;
+public:
+
+    static constexpr unsigned RANGES = log2floor(Max/Min);
+    static constexpr unsigned RESOLUTION = (NumBuckets - 1)/RANGES;
+    static constexpr unsigned RESOLUTION_BITS = log2floor(RESOLUTION);
+    static constexpr unsigned BASESHIFT = (Min == 0) ? 0 : log2floor(Min);
+    static constexpr uint64_t LOWER_BITS_MASK = (1 << RESOLUTION_BITS) - 1;
+    static_assert(BASESHIFT >= RESOLUTION);
+
+    approx_exponential_histogram() {
+        clear();
+    }
+
+    /*!
+     * \brief Returns the bucket lower limit given the bucket id.
+     * The first and last bucket will always return the MIN and MAX respectively.
+     *
+     */
+    uint64_t get_bucket_lower_limit(uint16_t bucket_id) const {
+        if (bucket_id == 0) {
+            return Min;
+        }
+        if (bucket_id == NumBuckets - 1) {
+            return Max;
+        }
+        int16_t range_id = (bucket_id >> RESOLUTION_BITS);
+        return (1 << (range_id + BASESHIFT)) +  ((bucket_id & LOWER_BITS_MASK) << (range_id + BASESHIFT - RESOLUTION_BITS));
+    }
+
+    /*!
+     * \brief Returns the bucket upper limit given the bucket id.
+     * The last bucket will return MAX.
+     *
+     */
+    uint64_t get_bucket_upper_limit(uint16_t bucket_id) const {
+        if (bucket_id == NumBuckets - 1) {
+            return Max;
+        }
+        return get_bucket_lower_limit(bucket_id + 1);
+    }
+
+    /*!
+     * \brief Find the bucket index for a given value
+     * The position of a value that is lower or equal to Min will always be 0.
+     * The position of a value is that is higher or equal to MAX will always be NUM_BUCKETS - 1.
+     */
+    uint16_t find_bucket_pos(uint64_t val) const {
+        if (val >= Max) {
+            return NumBuckets - 1;
+        }
+        if (val <= Min) {
+            return 0;
+        }
+        uint16_t range = log2floor(val);
+        val >>= range - RESOLUTION_BITS; // leave the top most N+1 bits where N is the resolution.
+        return ((range - BASESHIFT) << RESOLUTION_BITS) + (val & LOWER_BITS_MASK);
```

Comment:
@amnonh shouldn't `(val & LOWER_BITS_MASK)` be taken before val is shifted right above?
(Ref https://github.com/scylladb/scylla/issues/3806#issuecomment-645803974)